gpu: generic: implement SYCL reduction

sgeor255 opened the pull request:
# Checklist

## General

- [X] Do all unit and benchdnn tests (`make test` and `make test_benchdnn_*`) pass locally for each commit?
- [X] Have you formatted the code using clang-format?

## Performance improvements

Rbiessy commented on src/gpu/generic/sycl/ref_reduction.cpp:
```diff
@@ -0,0 +1,479 @@
+            return status::unimplemented;
+        }
+        if (needs_atomic_reduction_ && attr()->post_ops_.find(dnnl_sum) != -1) {
+            return status::unimplemented;
```
It seems there are a few cases where we would return `unimplemented`. Could we fallback to a simple implementation instead?

sgeor255 replied:
I have added a fallback simple implementation in #2249 .

Rbiessy commented on src/gpu/generic/sycl/ref_reduction.cpp:
```diff
@@ -0,0 +1,479 @@
+    }
+
+    if (needs_reorder_) { CHECK(init_reorder(engine)); }
+    if (needs_atomic_reduction_) { CHECK(init_out_scratchpad()); }
```
Can we check that the SYCL device supports atomic when `needs_atomic_reduction_` is `true`? I think we just need to check the query `info::device::atomic_memory_scope_capabilities` contains `memory_scope::work_group`.

sgeor255 replied:
Good point, done.

mgouicem commented on src/gpu/generic/sycl/ref_reduction.hpp:
```diff
@@ -0,0 +1,135 @@
+                    && attr()->has_default_values(sm::post_ops)
+                    && sycl_post_ops_t::post_ops_ok(attr())
+                    && attr_.set_default_formats(dst_md()) == status::success
+                    && src_wrap.is_plain() && dst_wrap.is_plain()
```
`is_plain` only checks that the memory is not blocked. 
However, below you seem to rely on the property that stride = accumulate(dims), which is not guaranteed by `is_plain`. Calling `is_dense` would guarantee that though.

sgeor255 replied:
Added `is_dense()` checks, thanks. I also kept the `is_plain()` checks as their main purpose was to ensure unimplemented is returned for blocked formats (which doesn't seem to be the case for `is_dense()`).

mgouicem reviewed (commented):
Overall, it seems to be a pretty optimized implementation for a reference one :).
I would suggest to at least have a baseline implementation with fewer restrictions (and maybe less efficient).

Few other comments:
- atomics based reduction can happen only under non deterministic mode (see [doc](https://oneapi-src.github.io/oneDNN/dev_guide_attributes_deterministic.html))
- sycl has a [reducer extension ](https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#sec:reduction)(mandatory per spec). Any reason to not rely on that?

mgouicem commented on src/gpu/generic/sycl/sycl_primitive_conf.hpp:
```diff
@@ -415,6 +415,39 @@ struct sycl_pooling_bwd_conf_t : public sycl_pooling_base_conf_t {
+
+    sycl_post_ops_t post_ops;
+
+    const int num_sg_reductions = 4;
+    static constexpr int local_row_wg = 256;
+    static constexpr int local_col_wg = 8;
```
Any reason for these values (maybe adding a few comments to clarify)?

sgeor255 replied:
These were derived from testing and are values that should work in most cases. I added a comment to actually tune them in the future when we do an optimisation pass.

sgeor255 commented on src/gpu/generic/sycl/reduction_kernels.hpp:
```diff
@@ -0,0 +1,658 @@
+    using Index = int;
+
+    static constexpr int Dim = 3;
+    static constexpr bool CheckBounds = true;
```
This is true, added the comment.

sgeor255 commented on src/gpu/generic/sycl/ref_reduction.cpp:
```diff
@@ -0,0 +1,479 @@
+} // namespace generic
+} // namespace gpu
+} // namespace impl
+} // namespace dnnl
```
Done.

sgeor255 commented:
@mgouicem 

> I would suggest to at least have a baseline implementation with fewer restrictions (and maybe less efficient).

Added a simple implementation in #2249 .

> * atomics based reduction can happen only under non deterministic mode (see [doc](https://oneapi-src.github.io/oneDNN/dev_guide_attributes_deterministic.html))

I had missed this, thanks. Added a check.

> * sycl has a [reducer extension ](https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#sec:reduction)(mandatory per spec). Any reason to not rely on that?

No particular reason. This implementation is based on an older implementation we had worked on before this extension was added.

> * test enabling: it seems gtest is disabled for GPU. It might be time to reenable it.

Enabled the tests. Also updated the documentation.

mgouicem commented on src/gpu/generic/sycl/ref_reduction.cpp:
```diff
@@ -0,0 +1,496 @@
+            return status::unimplemented;
+        }
+        if (needs_atomic_reduction_ && conf.alg == dnnl_reduction_mul) {
+            return status::unimplemented;
```
(minor) for these kind of checks, there is a `VDISPATCH` macro that allows to print diagnostics with ONEDNN_VERBOSE mode (see an example [here](https://github.com/oneapi-src/oneDNN/blob/0bff5bd0f56dcf5e68303198996e15a065ffb16d/src/gpu/generic/sycl/ref_convolution.hpp#L91)).

sgeor255 replied:
Done, thanks.

Rbiessy replied:
I pushed the change for him as Svet is on holiday

dzarukin commented on src/gpu/generic/sycl/reduction_kernels.hpp:
```diff
@@ -0,0 +1,659 @@
+
+struct init_kernel_t {
+    xpu::sycl::out_memory_arg_t out_arg_;
+    dnnl_alg_kind_t alg_;
```
Please replace with `alg_kind_t`. Please update all algorithms with `alg_kind::name`. They are meant to be used from an `alg_kind` namespace, not the public API C type.

Rbiessy replied:
This is done. I have applied similar changes to avoid using `dnnl_data_type_t` and `dnnl_dim_t`.

dzarukin commented on src/gpu/generic/sycl/reduction_kernels.hpp:
```diff
@@ -0,0 +1,659 @@
+namespace sycl {
+
+struct init_kernel_t {
+    xpu::sycl::out_memory_arg_t out_arg_;
```
Move members to private since there's a constructor?

dzarukin commented on src/gpu/generic/sycl/reduction_kernels.hpp:
```diff
@@ -0,0 +1,659 @@
+    sycl_post_ops_t post_ops_;
+    post_op_input_args po_args_;
+    xpu::sycl::md_t dst_md_;
+    int reduce_size_;
```
Same?

sgeor255 replied:
Done, thanks.

dzarukin commented on src/gpu/generic/sycl/reduction_kernels.hpp:
```diff
@@ -0,0 +1,659 @@
+    int col_lim_;
+    int local_id_;
+    int local_row_id_;
+    int local_col_id_;
```
`Index` type?

sgeor255 replied:
Done, thanks.

dzarukin commented on tests/gtests/test_reduction.cpp:
```diff
@@ -61,11 +59,22 @@ class reduction_test_t
     }
 
+    bool generic_supported_format_tag(memory::format_tag tag) {
+        return impl::utils::one_of(tag, dnnl_a, dnnl_ab, dnnl_abc, dnnl_abcd,
```
```suggestion
        return impl::utils::one_of(tag, format_tag::a, ...
```

sgeor255 replied:
Done, thanks.

AD2605 commented on src/gpu/generic/sycl/ref_reduction.cpp:
```diff
@@ -0,0 +1,499 @@
+        VDISPATCH_REDUCTION(
+                IMPLICATION(needs_atomic_reduction_, !attr()->deterministic_),
+                "Atomic reduction is only supported in non-deterministic mode");
+        VDISPATCH_REDUCTION(IMPLICATION(needs_atomic_reduction_,
+                                    conf.alg != alg_kind::reduction_mul),
+                "Algorithm Mul is not supported with atomic reduction");
```
IIUC, you are accumulating and storing it in f32, so you could implement an atomic multiplication using `compare_exchange_strong` and a do-while loop, which would enable any atomic operation we need.

sgeor255 replied:
As we discussed offline this case is handled by the fallback kernel and not that important to add currently.

AD2605 replied:
I am okay with it either way. It was more about the reducing the number of kernels rather than performance since it's a reference implementation anyway

sgeor255 commented on src/gpu/generic/sycl/ref_reduction.cpp:
```diff
@@ -0,0 +1,499 @@
+                                    conf.alg != alg_kind::reduction_mul),
+                "Algorithm Mul is not supported with atomic reduction");
+        VDISPATCH_REDUCTION(IMPLICATION(needs_atomic_reduction_,
+                                    attr()->post_ops_.find(dnnl_sum) == -1),
+                "Sum postop is not supported with atomic reduction");
```
The sum post-op requires the previous values stored in the output but with atomic reduction we do some intermediate computations and store the result in the output. This case is also handled by the fallback kernel.

AD2605 replied:
> The sum post-op requires the previous values stored in the output but with atomic reduction we do some intermediate computations and store the result in the output.

this needn't block the PR, but why cannot we add it to the final value of the reduction at the end, irrespective of the reduction type (atomic or not), to the output, instead of relying on a separate kernel ?

ShanoToni replied:
The main problem with atomic reduction and the post ops sum comes from the fact that in the atomic reduction case we are reducing a dimension which does not fit inside the workgroup. In this case multiple workgroups are used and we lose any kind of synchronization the only guarantee we have here is that the atomic operation will increment the value each work item in the different workgroups will accumulate without a race condition, we do not know which workgroup's work item would work in the data first. 

The only way i can see this being done is selecting only one workgroup's work items (for example wg 0) to apply the post-op sum with an atomic_add, which would require a custom way for reductions to apply the sum post op.

sgeor255 commented:
The CI failures seem unrelated to this PR (copyright checks failed for files in `src/cpu/` not added in this PR).

sgeor255 commented on src/gpu/generic/sycl/reduction_kernels.hpp:
```diff
@@ -0,0 +1,661 @@
+
+    template <::sycl::memory_order Order, ::sycl::memory_scope Scope,
+            ::sycl::access::address_space
+                    Space /* , typename InMem, typename OutMem */>
```
Done, thanks.

vpirogov commented:
@sgeor255, this change breaks gtests on Intel GPUs:
```
119/225 Test #119: test_reduction_gpu .......................................***Exception: Numerical  0.35 sec
Note: Google Test filter = *:-*_CPU*
[==========] Running 58 tests from 16 test suites.
[----------] Global test environment set-up.
[----------] 5 tests from TestReductionEF/reduction_test_f32
[ RUN      ] TestReductionEF/reduction_test_f32.TestsReduction/0
[       OK ] TestReductionEF/reduction_test_f32.TestsReduction/0 (0 ms)
[ RUN      ] TestReductionEF/reduction_test_f32.TestsReduction/1
[       OK ] TestReductionEF/reduction_test_f32.TestsReduction/1 (0 ms)
[ RUN      ] TestReductionEF/reduction_test_f32.TestsReduction/2
[       OK ] TestReductionEF/reduction_test_f32.TestsReduction/2 (0 ms)
[ RUN      ] TestReductionEF/reduction_test_f32.TestsReduction/3
[       OK ] TestReductionEF/reduction_test_f32.TestsReduction/3 (0 ms)
[ RUN      ] TestReductionEF/reduction_test_f32.TestsReduction/4
[       OK ] TestReductionEF/reduction_test_f32.TestsReduction/4 (0 ms)
[----------] 5 tests from TestReductionEF/reduction_test_f32 (0 ms total)

[----------] 1 test from TestReductionZero/reduction_test_f32
[ RUN      ] TestReductionZero/reduction_test_f32.TestsReduction/0

        Start 120: test_reduction_buffer_gpu
120/225 Test #120: test_reduction_buffer_gpu ................................***Exception: Numerical  0.35 sec
Note: Google Test filter = *:-*_CPU*
[==========] Running 58 tests from 16 test suites.
[----------] Global test environment set-up.
[----------] 5 tests from TestReductionEF/reduction_test_f32
[ RUN      ] TestReductionEF/reduction_test_f32.TestsReduction/0
[       OK ] TestReductionEF/reduction_test_f32.TestsReduction/0 (0 ms)
[ RUN      ] TestReductionEF/reduction_test_f32.TestsReduction/1
[       OK ] TestReductionEF/reduction_test_f32.TestsReduction/1 (0 ms)
[ RUN      ] TestReductionEF/reduction_test_f32.TestsReduction/2
[       OK ] TestReductionEF/reduction_test_f32.TestsReduction/2 (0 ms)
[ RUN      ] TestReductionEF/reduction_test_f32.TestsReduction/3
[       OK ] TestReductionEF/reduction_test_f32.TestsReduction/3 (0 ms)
[ RUN      ] TestReductionEF/reduction_test_f32.TestsReduction/4
[       OK ] TestReductionEF/reduction_test_f32.TestsReduction/4 (0 ms)
[----------] 5 tests from TestReductionEF/reduction_test_f32 (0 ms total)

[----------] 1 test from TestReductionZero/reduction_test_f32
[ RUN      ] TestReductionZero/reduction_test_f32.TestsReduction/0

```

sgeor255 commented:
> @sgeor255, this change breaks gtests on Intel GPUs:
> 
> ```
> 119/225 Test #119: test_reduction_gpu .......................................***Exception: Numerical  0.35 sec
> Note: Google Test filter = *:-*_CPU*
> [==========] Running 58 tests from 16 test suites.
> [----------] Global test environment set-up.
> [----------] 5 tests from TestReductionEF/reduction_test_f32
> [ RUN      ] TestReductionEF/reduction_test_f32.TestsReduction/0
> [       OK ] TestReductionEF/reduction_test_f32.TestsReduction/0 (0 ms)
> [ RUN      ] TestReductionEF/reduction_test_f32.TestsReduction/1
> [       OK ] TestReductionEF/reduction_test_f32.TestsReduction/1 (0 ms)
> [ RUN      ] TestReductionEF/reduction_test_f32.TestsReduction/2
> [       OK ] TestReductionEF/reduction_test_f32.TestsReduction/2 (0 ms)
> [ RUN      ] TestReductionEF/reduction_test_f32.TestsReduction/3
> [       OK ] TestReductionEF/reduction_test_f32.TestsReduction/3 (0 ms)
> [ RUN      ] TestReductionEF/reduction_test_f32.TestsReduction/4
> [       OK ] TestReductionEF/reduction_test_f32.TestsReduction/4 (0 ms)
> [----------] 5 tests from TestReductionEF/reduction_test_f32 (0 ms total)
> 
> [----------] 1 test from TestReductionZero/reduction_test_f32
> [ RUN      ] TestReductionZero/reduction_test_f32.TestsReduction/0
> 
>         Start 120: test_reduction_buffer_gpu
> 120/225 Test #120: test_reduction_buffer_gpu ................................***Exception: Numerical  0.35 sec
> Note: Google Test filter = *:-*_CPU*
> [==========] Running 58 tests from 16 test suites.
> [----------] Global test environment set-up.
> [----------] 5 tests from TestReductionEF/reduction_test_f32
> [ RUN      ] TestReductionEF/reduction_test_f32.TestsReduction/0
> [       OK ] TestReductionEF/reduction_test_f32.TestsReduction/0 (0 ms)
> [ RUN      ] TestReductionEF/reduction_test_f32.TestsReduction/1
> [       OK ] TestReductionEF/reduction_test_f32.TestsReduction/1 (0 ms)
> [ RUN      ] TestReductionEF/reduction_test_f32.TestsReduction/2
> [       OK ] TestReductionEF/reduction_test_f32.TestsReduction/2 (0 ms)
> [ RUN      ] TestReductionEF/reduction_test_f32.TestsReduction/3
> [       OK ] TestReductionEF/reduction_test_f32.TestsReduction/3 (0 ms)
> [ RUN      ] TestReductionEF/reduction_test_f32.TestsReduction/4
> [       OK ] TestReductionEF/reduction_test_f32.TestsReduction/4 (0 ms)
> [----------] 5 tests from TestReductionEF/reduction_test_f32 (0 ms total)
> 
> [----------] 1 test from TestReductionZero/reduction_test_f32
> [ RUN      ] TestReductionZero/reduction_test_f32.TestsReduction/0
> ```

Thanks for letting me know @vpirogov , I will take a look.

vpirogov commented:
It's a bit surprising that OpenCL runtime cases are broken too. This indicates the issue is coming from GPU common code, not from the implementation itself.

sgeor255 commented:
> It's a bit surprising that OpenCL runtime cases are broken too. This indicates the issue is coming from GPU common code, not from the implementation itself.

@vpirogov https://github.com/oneapi-src/oneDNN/pull/2453 should fix this.
